File: code_challenges/array_binary_search/array_binary_search.py
def mid_point (arr, index, key, lL, hL):
  index = (lL + hL) // 2

  if (arr[index] == key): return index
  elif (lL > hL): return -1
  elif (arr[index] < key): return mid_point(arr, index, key, index + 1, hL)
  else: return mid_point(arr, index, key, lL, index - 1)

def binary_search (arr, key):
  if not arr: return -1
  index, lower_limit, higher_limit = 0, 0, len(arr) - 1
  return mid_point(arr, index, key, lower_limit, higher_limit)

File: code_challenges/401-JS/arrayBinarySearch/array-binary-search.js
'use strict'

function BinarySearch (arr, key) {
  const index = Math.floor(arr.length / 2) - 1
  const lowerLimit = 0
  const higherLimit = arr.length - 1

  return midPoint(arr, index, key, lowerLimit, higherLimit)
}

function midPoint (arr, index, key, lL, hL) {
  while (true) {
    index = Math.floor((lL + hL) / 2)

    if (arr[index] === key) {
      return index
    } else if (lL > hL) {
      index = -1
      return index
    } else if (arr[index] < key) {
      lL = index + 1
    } else {
      hL = index - 1
    }
  }
}

function generateArray (length) {
  const arr = []
  for (let i = 0; i < length; i++) {
    arr[i] = Math.round(Math.random() * 100)
  }
  arr.sort((a, b) => a - b)
  const key = Math.round(Math.random() * 100)
  return [arr, key]
}
//
// let value = (generateArray(100));
// console.log(BinarySearch(value[0], value[1]));

module.exports = {
  BinarySearch: BinarySearch,
  generateArray: generateArray
}
